x-pack/plugin/mapper-aggregate-metric/src/main/java/org/elasticsearch/xpack/aggregatemetric/mapper/AggregateDoubleMetricFieldMapper.java

```diff
@@ -508,6 +511,107 @@ public ValueFetcher valueFetcher(SearchExecutionContext context, String format)
+            NumberFieldMapper.NumberFieldType minFieldType = metricFields.get(Metric.min);
+            NumberFieldMapper.NumberFieldType maxFieldType = metricFields.get(Metric.max);
+            NumberFieldMapper.NumberFieldType sumFieldType = metricFields.get(Metric.sum);
+            NumberFieldMapper.NumberFieldType countFieldType = metricFields.get(Metric.value_count);
+
+            private AggregateMetricDoubleBlockLoader() {}
+
+            static NumericDocValues getNumericDocValues(NumberFieldMapper.NumberFieldType field, LeafReader leafReader) throws IOException {
+                if (field == null) {
+                    return null;
+                }
+                String fieldName = field.name();
+                var values = leafReader.getNumericDocValues(fieldName);
+                if (values != null) {
+                    return values;
+                }
+
+                var sortedValues = leafReader.getSortedNumericDocValues(fieldName);
+                return DocValues.unwrapSingleton(sortedValues);
+            }
+
+            @Override
+            public AllReader reader(LeafReaderContext context) throws IOException {
+                NumericDocValues minValues = getNumericDocValues(minFieldType, context.reader());
+                NumericDocValues maxValues = getNumericDocValues(maxFieldType, context.reader());
+                NumericDocValues sumValues = getNumericDocValues(sumFieldType, context.reader());
+                NumericDocValues valueCountValues = getNumericDocValues(countFieldType, context.reader());
+
+                if (minValues == null || maxValues == null || sumValues == null || valueCountValues == null) {
+                    throw new UnsupportedOperationException("Must have all subfields to use aggregate double metric in ESQL");
+                }
+                return new BlockDocValuesReader() {
+
+                    private int docID = -1;
+
+                    @Override
+                    protected int docId() {
+                        return docID;
+                    }
+
+                    @Override
+                    public String toString() {
+                        return "BlockDocValuesReader.AggregateMetricDouble";
+                    }
+
+                    @Override
+                    public Block read(BlockFactory factory, Docs docs) throws IOException {
+                        try (var builder = factory.aggregateMetricDoubleBuilder(docs.count())) {
+                            int lastDoc = -1;
+                            for (int i = 0; i < docs.count(); i++) {
+                                int doc = docs.get(i);
+                                if (doc < lastDoc) {
+                                    throw new IllegalStateException("docs within same block must be in order");
+                                }
+                                read(doc, builder);
```

Comment:
My *instinct* is that we should read column at a time in this case - with four loops. Because the data is in column order. We can't in the row-style read method below, but this method wants to load column at a time so I'd do it. It may have performance benefits because you are packing into the same array rather than jumping around.

Reply:
Or those performance benefits may vanish into noise. But still, this method wants columns. Begs for it.